### PR TITLE
add ESM exports in package.json

### PR DESCRIPTION
**Pull request by akphi**
@krisk  We're using `Typescript` `module: Node16` in our project and since `Fuse.js` ESM does not specify `type: module` and `exports` field in `package.json`, it does not get picked up properly by Typescript. So I would like to add these to the `package.json` file.

The error I got is something like this, for the following import:

```ts
import Fuse from 'fuse.js';

const searchEngine = new Fuse(...

^
This expression is not constructable.
  Type 'typeof import("/Users/user/my-project/node_modules/fuse.js/dist/fuse")' has no construct signatures.ts(2351)
```

I believe this would help with https://github.com/krisk/Fuse/issues/541 as I have noted there

For Typescript esm mode, such as `Node16`, `NodeNext`, `esModuleInterop` is not [supported in](https://github.com/lukeed/clsx/issues/43) [this mode](https://github.com/microsoft/TypeScript/issues/49567#issuecomment-1157139107), it's worth if we do something like https://github.com/lukeed/clsx/pull/44 - I have given it an attempt but not quite sure how I can modify the `d.ts` file accordingly.

### CODE REVIEW

File: package.json
Comment (Jack-Works) [inline package.json:18]:
```suggestion
      "import": {
        "types": "./dist/fuse.d.ts",
        "default": "./dist/fuse.esm.js"
      },
      "require":  {
        "types": "./dist/fuse.d.ts",
        "default": "./dist/fuse.common.js"
      }
```
see https://github.com/microsoft/TypeScript/issues/50466

**Comment by rchandu** [comment]
This would be something that would help us. Is there something I can do to help here on this effort ?

**Comment by akphi** [comment]
@rchandu seems like there hasn't been active development in this repo for a couple of months now 🤔

File: package.json
Comment (favna) [inline package.json:20]:
```suggestion
  "exports": {
    "types": "./dist/fuse.d.ts",
    "import": "./dist/fuse.esm.js",
    "require":  "./dist/fuse.common.js"
  },
```

The old version was invalid. You cannot just add `"type": "module"` unless Fuse itself is rewritten to ESM. Also `"type": "module"` only affects the module itself, not when the module is loaded through node_modules.

Furthermore the subgrouping of `"."` is not necessary because Fuse doesn't have subpath exports

Lastly, after including this change thus reverting `"type": "module"` you should also revert the other `.cjs` related changes.

**Review by favna — CHANGES_REQUESTED** [review]

File: package.json
Comment (akphi) [reply package.json]: updated! Thanks for reviewing!

File: package.json
Comment (akphi) [reply package.json]: Thanks for reviewing! I have updated to your suggestions.

It's been a while and I haven't checked again, locally, but I hope this would work well with Typescript + Jest

**Comment by akphi** [comment]
@krisk Could you kindly please take a look at this? Thanks!

File: package.json
Comment (favna) [reply package.json]: You'll also have to rebase it because I dropped Jest in #711 and the repo now uses Vitest

Comment (favna) [reply package.json]: It should. I maintain the [Sapphire project](https://github.com/sapphiredev) and we use this structure in all projects there are it works with a variety of stacks. CJS, ESM, Jest, Vitest, etc

**Comment by favna** [comment]
@akphi expect a PR to your fork with a completely refreshed but also thoroughly tested setup very soon, when you merge that it'll pop up here for Krisk to then review fully.

I noticed my comment above introduced some other issues as well but now I am testing the scenarios:
- CJS (no fluff, super basic package.json)
- MJS (with type module and/or mjs file extension)
- TypeScript with `"module": "Node16"` and "moduleResolution": "Node"` (essentially modern CJS)
- TypeScript with `"module": "Node16"` and "moduleResolution": "Node16"` (essentially modern ESM)

**Comment by favna** [comment]
@akphi I created https://github.com/akphi/Fuse/pull/1 just now but before it can be merged you need to rebase your own `exports` branch on top of `Krisk/master`. Alternatively I _could_ create my own PR from my own fork and we close this PR but that means your contribution credit is lost, your call.

**Comment by akphi** [comment]
hey @favna, thanks for being super transparent; I really don't mind you carrying that out; all I need is just to have ESM ready :) Thanks so much for the help!

**Comment by favna** [comment]
I have created https://github.com/krisk/Fuse/pull/727 so this PR can be closed @akphi

**Pull request closed without merging**
